File: R/Scripts/statistical_experiments/permutation_testing.R
library(ggplot2)
data <- read.csv("Data/web_page_data.csv")

ggplot(data, aes(x=Page, Y=Time)) +
  geom_boxplot()

# compare means
mean_a <- mean(data[data["Page"] == "Page A", "Time"])
mean_b <- mean(data[data["Page"] == "Page B", "Time"])

mean_b - mean_a

permutation_func <- function(x, nA, nB) {
  n <-nA + nB
  idx_b <- sample(1:n, nB)
  idx_a <- setdiff(1:n, idx_b)
  mean_diff <- mean(x[idx_b]) - mean(x[idx_a])
  return(mean_diff)
}

perm_diffs <- rep(0,1000)
for (i in 1:1000) {
  perm_diffs[i] = permutation_func(data[, "Time"],21,15)
}
hist(perm_diffs, xlab =  "Session time differences")
abline(v=mean_b - mean_a, col="red", lwd = 3)

# how often mean difference of random permutation exceeds observed difference
mean(perm_diffs > (mean_b - mean_a))

#observed differences in session time between page A and B is well within the range of chance variation
#thus in not statistically significant 

#histogram of random;y permuted differences in conversion

obs_pct_diff <- 100 * (200 / 23739 - 182 / 22588)
conversion <- c(rep(0, 45945), rep(1, 382))
perm_diffs <- rep(0, 1000)
for (i in 1:1000) {
  perm_diffs[i] = 100 * permutation_func(conversion, 23739, 22588)
}
hist(perm_diffs)
abline(v=obs_pct_diff,lty=2, lwd=3,col="red")

#p-value
mean(perm_diffs > obs_pct_diff) # from permutation test
prop.test(x=c(200,182), n=c(23739,22588), alternative='greater') #p-value approximation (can use for binomial distribution)

#t-Test
t.test(Time ~ Page, data = data)
t.test(Time ~ Page, data = data, alternative ='less')

File: Python/statistics/statistical_experiments/permutation_testing.py
from random import sample

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


def permutation(x: pd, nA: int, nB: int):
    n = nA + nB
    idx_b = set(sample(range(n), nB))
    idx_a = set(range(n)) - idx_b
    return x.loc[idx_b].mean() - x.loc[idx_a].mean()


if __name__ == "__main__":
    file = "../../data/statistics/web_page_data.csv"
    data = pd.read_csv(file)

    data.boxplot(by="Page", column="Time")
    plt.show()

    # compare means
    mean_a = data[data.Page == "Page A"].Time.mean()
    mean_b = data[data.Page == "Page B"].Time.mean()

    print(f"mean diff: {mean_a - mean_b}")

    nA = data[data.Page == 'Page A'].shape[0]
    nB = data[data.Page == 'Page B'].shape[0]

    perm_diff = [permutation(data.Time, nA, nB) for _ in range(1000)]

    fig, ax = plt.subplots(figsize=(5, 5))
    ax.hist(perm_diff, bins=11, rwidth=0.9)
    ax.axvline(x=mean_b - mean_a, color="red", lw=3)
    ax.text(50,190, "Observed\ndifference",
            bbox={"facecolor":"blue"})
    ax.set_xlabel("Session time differences")
    ax.set_ylabel("Frequency")

    plt.show()

    # how often mean difference of random permutation exceeds observed difference
    print(np.mean(perm_diff > mean_b - mean_a))
    """observed differences in session time between page A and B is well within the range of chance variation
    thus in not statistically significant 
    """